src/events/input_provider.rs

```diff
@@ -0,0 +1,254 @@
+//! Contains the `InputProvider` trait, which is used to provide input events to widgets.
+
+use events::{UiEvent, Scroll, MouseClick, MouseDrag, InputState};
+use input::{Input, Button};
+use input::keyboard::Key;
+use input::mouse::MouseButton;
+use position::Point;
+
+
+/// Trait for something that provides events to be consumed by a widget.
+/// Provides a bunch of convenience methods for filtering out specific types of events.
+pub trait InputProvider<'a, T: Iterator<Item=&'a UiEvent>> {
+    /// This is the only method that needs to be implemented.
+    /// Just provided a reference to a `Vec<UiEvent>` that contains
+    /// all the events for this update cycle.
+    fn all_events(&'a self) -> T;
+
+    /// Returns the current input state. The returned state is assumed to be up to
+    /// date with all of the events so far.
+    fn current_state(&'a self) -> &'a InputState;
+
+    /// If the given mouse button is currently pressed, returns the current position of the mouse.
+    /// Otherwise, returns `None`
+    fn mouse_button_down(&'a self, button: MouseButton) -> Option<Point>;
+
+    //////////////////////////////////////////////////
+    // Methods that just check the stream of events //
+    //////////////////////////////////////////////////
+
+    /// Returns a `String` containing _all_ the text that was entered since
+    /// the last update cycle.
+    fn text_just_entered(&'a self) -> Option<String> {
```

Comment:
I think this is a nice convenience method for users who aren't too concerned about allocations, however I think we should also offer a non-allocating version of this method too - one that instead returns some Iterator yielding `&str`s.